src/device/clint.rs

```diff
@@ -154,20 +254,43 @@ impl VirtClint {
                     todo!("Setting mtime for another hart is not yet supported");
                 }
 
-                // Update the virtual `mip` according to the relative ordering of mtime and
-                // mtimecmp.
-                if mtime >= value {
-                    ctx.csr.mip |= mie::MTIE_FILTER;
+                // When we receive a timer, we clear the corresponding interrupt bit
+                if is_from_payload {
+                    self.next_timestamp_payload[ctx.hart_id]
+                        .counter
+                        .store(value, Ordering::SeqCst);
+                    ctx.csr.mip &= !mie::STIE_FILTER;
                 } else {
-                    // Register a timer to trigger the virtual interrupt once appropriate
-                    driver.write_mtimecmp(hart, value)?;
+                    self.next_timestamp_firmware[ctx.hart_id]
+                        .counter
+                        .store(value, Ordering::SeqCst);
                     ctx.csr.mip &= !mie::MTIE_FILTER;
                 }
 
+                let is_interrupt_ready: bool = mtime >= value;
+
+                if is_interrupt_ready {
+                    ctx.csr.mip |= if is_from_payload {
+                        mie::STIE_FILTER
+                    } else {
+                        mie::MTIE_FILTER
+                    };
+                } else {
+                    let mtimecmp_firmware = self.next_timestamp_firmware[ctx.hart_id]
+                        .counter
+                        .load(Ordering::SeqCst);
+                    let mtimecmp_payload = self.next_timestamp_payload[ctx.hart_id]
+                        .counter
+                        .load(Ordering::SeqCst);
+
+                    driver.write_mtimecmp(hart, min(mtimecmp_firmware, mtimecmp_payload))?;
+                }
+
+                self.propagate_payload_interupt_physically(ctx);
```

Comment:
We should do that only when updating the clint of behalf of the payload. The reason is that if we update the `mip.STIP` bit while the firmware is running then we will get and supervisor interrupt while running the firmware, which we definitely don't want :)

It might maybe not happen with the code right now, but let's be defensive and add an if there just to limit the risk of introducing bugs with future refactorings.